tests/storages_tests/rdb_tests/test_storage.py

```diff
@@ -177,6 +184,82 @@ def test_upgrade() -> None:
+        storage_url = f"sqlite:///{workdir}/sqlite.db"
+
+        storage = RDBStorage(storage_url, skip_compatibility_check=True)
+        assert storage.get_current_version() == f"v{optuna_version}.a"
+        head_version = storage.get_head_version()
+        storage.upgrade()
+        assert head_version == storage.get_current_version()
+
+        # Create a new study.
+        study = create_study(storage=storage)
+        assert len(study.trials) == 0
+        study.optimize(objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 1
+
+        # Check empty study.
+        study = load_study(storage=storage, study_name="single_empty")
+        assert len(study.trials) == 0
+        study.optimize(objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 1
+
+        # Resume single objective optimization.
+        study = load_study(storage=storage, study_name="single")
+        assert len(study.trials) == 1
+        study.optimize(objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 2
+        for trial in study.trials:
+            assert trial.system_attrs["a"] == 0
+            assert trial.user_attrs["b"] == 1
+            assert trial.intermediate_values[0] == 0.5
+            assert -5 <= trial.params["x"] <= 5
+            assert 0 <= trial.params["y"] <= 10
+            assert trial.params["z"] in (-5, 0, 5)
+            assert trial.value is not None and 0 <= trial.value <= 150
+
+        assert study.system_attrs["c"] == 2
+        assert study.user_attrs["d"] == 3
+
+        # optuna<2.4.x does not support multi-objective optimization officially, so skip them.
+        if packaging.version.parse(optuna_version) < packaging.version.parse("2.4.0"):
+            return
+        # Create a new study.
+        study = create_study(storage=storage, directions=["minimize", "minimize"])
+        assert len(study.trials) == 0
+        study.optimize(mo_objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 1
+
+        # Check empty study.
+        study = load_study(storage=storage, study_name="multi_empty")
+        assert len(study.trials) == 0
+        study.optimize(mo_objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 1
+
+        # Resume multi-objective optimization.
+        study = load_study(storage=storage, study_name="multi")
+        assert len(study.trials) == 1
+        study.optimize(mo_objective_test_upgrade, n_trials=1)
+        assert len(study.trials) == 2
+        for trial in study.trials:
+            assert trial.system_attrs["a"] == 0
+            assert trial.user_attrs["b"] == 1
+            assert -5 <= trial.params["x"] <= 5
+            assert 0 <= trial.params["y"] <= 10
+            assert trial.params["z"] in (-5, 0, 5)
+            assert -5 <= trial.values[0] < 5
+            assert 0 <= trial.values[1] <= 150
+        assert study.system_attrs["c"] == 2
+        assert study.user_attrs["d"] == 3
```

Comment:
How about splitting the test suite for these scenarios? It looks like a code smell of [conditional test logic](http://xunitpatterns.com/Conditional%20Test%20Logic.html).

Reply:
That makes sense. I split `test_upgrade` into `test_upgrade_single_objective_optimization` and `test_upgrade_multi_objective_optimization` in https://github.com/optuna/optuna/pull/2890/commits/a36e0538b9067a4f542fc240621de58ad67d0171. Thank you!